### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by antondalgren**
Rewrite to utilize Typescript.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -9,6 +9,10 @@
       "sourceType": "module"
     },
     "rules": {
-        "no-console": "off"
-    }
+      "no-console": "off",
+      "semi": ["error", "never"],
+      "no-unused-vars": "off"
```
Comment (carlhoerberg) [inline .eslintrc.json:14]: why?

**Review by carlhoerberg — CHANGES_REQUESTED** [review]
Can `dist/amqp-websocket-client.mjs` be generated with 2 space indentation instead of 4? then the diff will be easier to review

File: .eslintrc.json (hunk 1 above)
Comment (antondalgren) [reply .eslintrc.json]: No actual reason. Changed it to Error :)

**Comment by carlhoerberg** [comment]
how about docs, can they still be generated?

**Comment by carlhoerberg** [comment]
While running `npm test` in this branch without a rabbitmq server running i get 17 failed and one `unhandled rejection`. while in the main branch i only get 17 failed, no unhandled rejections.

**Comment by carlhoerberg** [comment]
also in main: 

```
 can publish and consume

  Rejected promise returned by test
```

this branch: 
```
can publish and consume

  Error: Promise returned by test never resolved

  › process.emit (node_modules/source-map-support/source-map-support.js:495:21)
```

**Comment by carlhoerberg** [comment]
Missing types on basicQos and basicPrefetch method arguments

**Comment by antondalgren** [comment]
> how about docs, can they still be generated?

Yes. Might need to tweak it, but i'd prefer to have this and #3 merged before, to not have to care about the dist files. 

> While running npm test in this branch without a rabbitmq server running i get 17 failed and one unhandled rejection. while in the main branch i only get 17 failed, no unhandled rejections.

Why would you want to run the test suite without RabbitMQ running?


> also in main:
> 
> ```
>  can publish and consume
> 
>   Rejected promise returned by test
> ```
> 
> this branch:
> 
> ```
> can publish and consume
> 
>   Error: Promise returned by test never resolved
> 
>   › process.emit (node_modules/source-map-support/source-map-support.js:495:21)
> ```

Is this with/without RabbitMQ running? I have 17 passes and 1 skipped on both main and this branch, CI also passes.

**Comment by antondalgren** [comment]
> Missing types on basicQos and basicPrefetch method arguments

The method arguments have types. If a default value, eg. `prefetchSize = 0` is set, the typescript compiler knows that this is of type `number`.

https://github.com/cloudamqp/amqp-client.js/pull/1/files#diff-663dae8bbcf9718ab8a7ac19e5296ecafa56b738b6c2d2e76a3214b24048b5daR54 

https://github.com/cloudamqp/amqp-client.js/pull/1/files#diff-663dae8bbcf9718ab8a7ac19e5296ecafa56b738b6c2d2e76a3214b24048b5daR394

**Comment by antondalgren** [comment]
> Missing types on basicQos and basicPrefetch method arguments


> also in main:
> 
> ```
>  can publish and consume
> 
>   Rejected promise returned by test
> ```
> 
> this branch:
> 
> ```
> can publish and consume
> 
>   Error: Promise returned by test never resolved
> 
>   › process.emit (node_modules/source-map-support/source-map-support.js:495:21)
> ```

Solved in: d9cf64c

**Comment by carlhoerberg** [comment]
i see, it seems like it's the documentation thingy that get's confused where it's `number?` and then converts to `any`

**Comment by antondalgren** [comment]
> i see, it seems like it's the documentation thingy that get's confused where it's `number?` and then converts to `any`

@carlhoerberg  Yeah, the hardened eslint config does not want me to write `prefetchSize: number = 0` as the type `number` is automatically inferred by the typescript compiler. But it seems like the documentation tool does not catch that. But i'd like to work on solving that after this and #3 is merged. Either by adding the type back to the JSDoc comments or if I manage to find something that figures it out like the compiler does.

**Comment by antondalgren** [comment]
FYI, I tried installing this branch using NPM and it worked as expected. 

```node
import AMQPClient from '@cloudamqp/amqp-client'


async function run() {
  try {
    const amqp = new AMQPClient("amqp://localhost")
    const conn = await amqp.connect()
    const ch = await conn.channel()
    const q = await ch.queue()
    const consumer = await q.subscribe({noAck: true}, async (msg) => {
      console.log(msg.bodyToString())
      await consumer.cancel()
    })
    await q.publish("Hello World", {deliveryMode: 2})
    await consumer.wait() // will block until consumer is canceled or throw an error if server closed channel/connection
    await conn.close()
  } catch (e) {
    console.error("ERROR", e)
    e.connection.close()
    setTimeout(run, 1000) // will try to reconnect in 1s
  }
}

run()
```

**Comment by carlhoerberg** [comment]
Added typescript definitions, but just by extending the jsdocs, in #12, but it was good, typescript caught a lot of subtle bugs we had

**Pull request closed without merging**